Add cmake build code

SchaichAlonso opened the pull request:
Preliminary PR to add a cmake project file.

Fixes: #59

SchaichAlonso committed:
Add some cmake build files

SchaichAlonso committed:
File used by tkrzw_dbm_perf _must_ have .tkh sufix

SchaichAlonso committed:
Multijob make safety++

Force serial execution of everything within check-hashdbm-util,
and rename `casket*` to `check-hashdbm-util*` in order to preempt
the delete statement from racing with the other tests.

SchaichAlonso committed:
Optionalize CLI application compilation and test execution

SchaichAlonso committed:
Autogenerate DLL symbol table

tkrzw doesn't explicitly set any symbol exports, therefore
relying on gcc's default policy to export everything.

On msvc, the default behaviour is to not export unless
explicitly asked for.

Have cmake instruct msvc to export everything, to match
the gcc behaviour

shenlebantongying commented on CMakeLists.txt:
```diff
@@ -0,0 +1,342 @@
+cmake_minimum_required(VERSION 3.28)
+
+cmake_policy(SET CMP0048 NEW)
+cmake_policy(SET CMP0091 NEW)
```
These two lines doesn't appear to be needed if cmake_minimum_required is 3.28?

SchaichAlonso replied:
Hi.

In cmake, most [policies](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Policies.html) default to the `OLD` behaviour, in order to remain compatible to previous behaviour when people bump up `cmake_mininum_required_version`.

[CMP0048](https://cmake.org/cmake/help/latest/policy/CMP0048.html)'s last paragraph states that unless explicitly set, the old behaviour will be used.
The same goes for [CMP0091](https://cmake.org/cmake/help/latest/policy/CMP0091.html).

shenlebantongying replied:
No?

[CMP0091](https://cmake.org/cmake/help/v3.28/policy/CMP0091.html) page 

> Use the [cmake_policy()](https://cmake.org/cmake/help/v3.28/command/cmake_policy.html#command:cmake_policy) command to set it to OLD or NEW explicitly.

The [cmake_minimum_required](https://cmake.org/cmake/help/v3.28/command/cmake_minimum_required.html#policy-settings) page from v3.28 says

> All policies known to the running version of CMake and introduced in the <min> (or <max>, if specified) version or earlier will be set to use NEW behavior.

I think this means `cmake_minimum_required(VERSION 3.28)` will set `NEW` for all the policies introduced in `VERSION <= 3.28`.

shenlebantongying commented on CMakeLists.txt:
```diff
@@ -0,0 +1,342 @@
+find_package(Threads)
+target_link_libraries(tkrzw PUBLIC Threads::Threads)
+
+# See https://gitlab.kitware.com/cmake/cmake/-/issues/18837
+if(MSVC)
+	target_compile_options(tkrzw PUBLIC "/Zc:__cplusplus")
+endif()
+
+target_compile_definitions(tkrzw
+    PRIVATE
+        _FILE_OFFSET_BITS=64
+        _TKRZW_PKG_VERSION="${CMAKE_PROJECT_VERSION}"
+        _TKRZW_LIB_VERSION="${LIB_VERSION}"
+)
+
+target_compile_features(tkrzw PUBLIC cxx_std_17)
```
No opinion about this line, this means if the compiler by default uses 20 or 23 then the code will be compiled in 20 or 23 mode.

Alternatively, set the version to fixed 17

```
set_property(TARGET tkrzw PROPERTY CXX_STANDARD 17)
```

shenlebantongying commented on CMakeLists.txt:
```diff
@@ -0,0 +1,342 @@
+)
+set_target_properties(tkrzw
+    PROPERTIES
+        MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug>$<IF:$<STREQUAL:$<TARGET_PROPERTY:tkrzw,TYPE>,SHARED_LIBRARY>,DLL,>
```
I don't have access to Windows machine now, but last time I checked, for relative newer CMake, this is useless. However, I am not sure at all.

SchaichAlonso replied:
It was not required for older versions of cmake, i.e. before 3.15.
Starting with 3.15, [CMP0091](https://cmake.org/cmake/help/latest/policy/CMP0091.html) deprecated the automatic setup of the correct runtime library.

If `CMP0091` isn't set at the prelude section, the explicit setting of [MSVC_RUNTIME_LIBRARY](https://cmake.org/cmake/help/latest/prop_tgt/MSVC_RUNTIME_LIBRARY.html) is not required. In my opinion, the old behaviour was "friendlier", as this line is more or less a triviality, but now needs to be written all over the place. The `OLD` behaviour, however, is deprecated.

shenlebantongying replied:
No?

[CMP0091](https://cmake.org/cmake/help/v3.28/policy/CMP0091.html) page from v3.28 says 

> Use the [cmake_policy()](https://cmake.org/cmake/help/v3.28/command/cmake_policy.html#command:cmake_policy) command to set it to OLD or NEW explicitly.


The [cmake_minimum_required](https://cmake.org/cmake/help/v3.28/command/cmake_minimum_required.html#policy-settings) page says

> All policies known to the running version of CMake and introduced in the <min> (or <max>, if specified) version or earlier will be set to use NEW behavior.

I think this means `cmake_minimum_required(VERSION 3.28)` will set `NEW` for all the policies introduced in `VERSION <= 3.28`.

shenlebantongying commented:
Impressive, I learned some new thing from this. I am not associated with this project in any way. Just a few nitpicking.

shenlebantongying commented on CMakeLists.txt:
```diff
@@ -0,0 +1,342 @@
+        _TKRZW_LIB_VERSION="${LIB_VERSION}"
+)
+
+target_compile_features(tkrzw PUBLIC cxx_std_17)
```
No opinion about this line, this means if the compiler by default uses 20 or 23 then the code will be compiled in 20 or 23 mode.

Alternatively, set the version to fixed 17

```
set_property(TARGET tkrzw PROPERTY CXX_STANDARD 17)
```

SchaichAlonso replied:
I agree with `CXX_STANDARD` being more readable.

However, it is ancient. and was/is meant as a relieve from the days people were writing hundreds of lines of cmake code to add the compiler specific option for `-std=c++11` (which might require to be denoted as `-std=c++0x`). Because all those self-written cxx_standard code had tweaks and fallbacks and odd hacks, `CXX_STANDARD` doesn't actually enforce anything unless [CXX_STANDARD_REQUIRED](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD_REQUIRED.html) is also set. And then, [CXX_EXTENSIONS](https://cmake.org/cmake/help/latest/prop_tgt/CXX_EXTENSIONS.html) has to be disabled, as the default to `CXX_STANDARD` would enable vendor specific extentions, which isn't what the autoconf code was doing.

All three options, unlike `cxx_std_XX`, can be set globaly, with something like

    set (CMAKE_CXX_STANDARD 14)
    set (CMAKE_CXX_EXTENSIONS OFF)
    set (CXX_STANDARD_REQUIRED ON)

 so the fact it's 3 options to set would be a minor issue for large projects with many targets.

However, `CXX_STANDARD` does not trancend, which is a showstopper: Inside of the `tkrzw` build, setting up c++17 mode, either globaly or by a lot of `set_property` calls, might be good enough to build `tkrzw`.

When the built targets gets installed, however, this [configure_package_config_file](https://github.com/SchaichAlonso/tkrzw/blob/3653a70a3a8d02d39cb2e087d057e4159ad86f2a/CMakeLists.txt#L160) call generates a file [that is then installed](https://github.com/SchaichAlonso/tkrzw/blob/3653a70a3a8d02d39cb2e087d057e4159ad86f2a/CMakeLists.txt#L175), containing everything _other_ projects require in order to use the tkrzw targets, similar to the [pkg-config file](https://github.com/estraier/tkrzw/blob/160e5cb618d80c540c05ef5e2bf4a9df9deb0d68/tkrzw.pc.in) that is generated by autotools.
The other project doesn't know anything about global variable setups in tkrzw, and as `CXX_STANDARD` doesn't transcend, "using" tkrzw will not imply the requirement for a c++17 mode to the consuming project. `cxx_std_XX` does transcends to all consumers, up- but not down-grading the c++ standard in use, i.e. if a consumer uses c++11 itself but links against tkrzw, it will be upgraded to use c++17, while a project using c++23 will keep using 23 despite linking against tkrzw.

I'm currently working on this integration.

shenlebantongying commented:
Just a few nitpicking. I am not associated with this project in any way. Overall looks good.

shenlebantongying commented on .github/workflows/cmake.yml:
```diff
@@ -0,0 +1,54 @@
+name: Build and Test
```
```suggestion
name: CMake Build and Test
```

shenlebantongying commented on CMakeLists.txt:
```diff
@@ -0,0 +1,342 @@
+include(GNUInstallDirs)
+include(CMakeDependentOption)
+
+option(WITH_TOOLS "Build cli frontends" OFF)
+cmake_dependent_option(WITH_TESTS "Generate a test target" OFF "WITH_TOOLS" OFF)
```
CMake has hidden feature that groups names based on prefix (Used by CMake GUI, Qt creator, VS2022...)

This should just be `TKRZW_TOOLS`, `TKRZW_TESTS`. 

They are grouped into `WITH`. I also used WITH in another two projects, and they cannot fix it anymore 😅 


<img width="200" src="https://github.com/user-attachments/assets/4d880c33-9792-460f-9d5e-d59d0810814f">

SchaichAlonso replied:
Indeed, a `TKRZW_` prefix scales better then a `WITH_` prefix does.

shenlebantongying commented:
Overall, looks good and I tested briefly locally. Just a few nitpicking. I am not associated with this project in any way as of now.

SchaichAlonso committed:
Rename github actions pipeline

Co-authored-by: shenleban tongying <shenlebantongying@gmail.com>

SchaichAlonso committed:
Prefix options with TKRZW_ rather then WITH_

SchaichAlonso commented:
VCPKG overlay repository: https://github.com/SchaichAlonso/tkrzw-vcpkg-overlay

SchaichAlonso commented:
VCPKG consumer example repository that builds the demos in `example` of the main repository against a vcpkg provisioned tkrzw package: https://github.com/SchaichAlonso/tkrzw-vcpkg-consumer-example

Seems [there's some assertion blowing up](https://github.com/SchaichAlonso/tkrzw-vcpkg-consumer-example/actions/runs/11707278969/job/32606291177#step:5:135) in `treedbm_ex1`

SchaichAlonso committed:
Move implementation out of interface headers (estraier/tkrzw#61)

tkrzw has quite a bit of code that is implemented in header files
that are supposed to be included by consuming projects.

This causes some technical issues with [weak symbols](https://en.wikipedia.org/wiki/Weak_symbol) inside
the consumer's translation units (or `.obj` files), that might fail
to be deduplicated if the consuming project uses compiler settings
different from those used by the tkrzw build system.

More importantly, however, parts of the code, especially imlementations
of non-virtual functions, might end up being embedded into consumer
applications when an optimization phase eliminates the procedure
calls, raising many concerns considering licensing.

This commit moves all implementations from header files documented
for inclusion into third-party projects on https://dbmx.net/tkrzw/,
as well as any header that implicitly gets included by including a
tkrzw interface header into source files.
Further, `inline` declarations are removed as within tkrzw, the
inlining happens on the discredition of the linker independed of
the `inline` keyword, while cross-module inlining is what this
commit is supposed to interdict.

Nothing can be done about template code in tkrzw_lib_common.h ,
however, template code is whitelisted on GPLv3.

SchaichAlonso committed:
<cmath> is needed for std::isnan()

SchaichAlonso committed:
Don't duplicate default arguments

SchaichAlonso committed:
Remove default argument from source file

SchaichAlonso committed:
Provide accessor functions for globals in tkrzw_lib_common

SchaichAlonso committed:
[ABI break] Runtime-detect page size on Windows

While ia32 and amd64 use 4096 byte pages in popular setups,
however all i686 processors have the ability to use 8k pages,
while Xeon E5/E7/Scalable have up to 1GB "Huge Page" support.

ia64 systems never had 4k page support to begin with and use
64k pages instead.

Autodetect the page size on win32 systems.

SchaichAlonso committed:
Use global accessors for global variables

SchaichAlonso committed:
Add and use accessors to global variables

SchaichAlonso committed:
Catch up to b0b14b3

SchaichAlonso committed:
Catch up to estraier/tkrzw#62

SchaichAlonso committed:
Build DLL variant

SchaichAlonso committed:
Name accessors to magic ids accordingly

It might be better to have a custom type with a real operator==
though

SchaichAlonso committed:
Add/use accessor to global variable

SchaichAlonso committed:
Add more accessors for global variables

SchaichAlonso committed:
Unit tests (#5)

Build and execute unit tests, with the exception of tkrzw_langc_test which has just too many global variable accesses to get them all.

SchaichAlonso committed:
Finish all CI slaves even if one fails

SchaichAlonso committed:
Rename stale identifier

SchaichAlonso committed:
[...]
